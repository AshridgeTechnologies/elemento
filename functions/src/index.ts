import * as functions from 'firebase-functions'
import axios from 'axios'
import express from 'express'

const app = express()

app.use(express.raw({type: 'application/zip'}))
app.use( (req, res, next) => {
    console.log(req.method, req.url)
    next()
})
app.put('/uploadfunctioncontent', async (req, res) => {
    try {
        const data = req.body
        const {path, GoogleAccessId, Expires, Signature} = req.query
        const url = `https://storage.googleapis.com${path}?GoogleAccessId=${GoogleAccessId}&Expires=${Expires}&Signature=${encodeURIComponent(Signature as string)}`
        console.log('upload url', url)

        await axios.put(url, data, {
            headers: {
                'Content-Type': req.headers['content-type'] as string
            }
        })
        res.end()
    } catch (e) {
        console.error(e)
        res.sendStatus(500)
    }
})

export const uploadfunctioncontent = functions.https.onRequest(app)
