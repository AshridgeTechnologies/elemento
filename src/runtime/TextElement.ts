import React from 'react'
import {Typography} from '@mui/material'

export default function TextElement({children}: {children: any}) {
    return React.createElement(Typography, {gutterBottom: true}, children)
}
