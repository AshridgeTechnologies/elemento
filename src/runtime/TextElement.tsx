import React from 'react'
import {Typography} from '@mui/material'
import {isArray} from 'lodash'
import {asText, valueOfProps} from './runtimeFunctions'

type Properties = {path: string, children?: any
    fontSize?: string | number, fontFamily?: string, color?: string, backgroundColor?: string,
    border?: number, borderColor?: string, width?: number, height?: number
    }

export default function TextElement({children, path, ...props}: Properties) {
    const childArray = isArray(children) ? children : [children]
    const propVals = valueOfProps(props)
    return React.createElement(Typography, {id: path, gutterBottom: true, ...propVals}, ...childArray.map(asText))
}
