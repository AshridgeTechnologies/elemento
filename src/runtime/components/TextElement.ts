import React from 'react'
import {Typography} from '@mui/material'
import lodash from 'lodash'; const {isArray, isFunction, isPlainObject, isObject, isString} = lodash;
import {PropVal, StylesProps, valueLiteral, valueOfProps} from '../runtimeFunctions'
import {flatten} from 'ramda'

type Properties = Readonly<{path: string, children?: any,
    show?: PropVal<boolean>,
    styles?: StylesProps
    }>

function asText(content: any) {
    if (React.isValidElement(content)) return content
    const contentValue = content?.valueOf()
    if (isString(contentValue) && contentValue.includes('\n')) {
        return flatten(contentValue.split('\n').map((line: string, index: number) => [line, React.createElement('br', {key: index})]))
    }
    if (isPlainObject(contentValue)) {
        return valueLiteral(contentValue)
    }
    if (isFunction(contentValue)) {
        return 'function ' + contentValue.name
    }
    if (isObject(contentValue)) {
        return contentValue.toString()
    }
    return contentValue
}

export default function TextElement({children, path, ...props}: Properties) {
    const childArray = isArray(children) ? children : [children]
    const {show = true, styles = {}} = valueOfProps(props)
    const showProps = show ? {} : {display: 'none'}
    const sxProps = {
        ...showProps,
        ...styles
    }

    return React.createElement(Typography, {id: path, sx: sxProps}, ...childArray.map(asText))
}
