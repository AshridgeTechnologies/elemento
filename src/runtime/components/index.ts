export {default as App} from './App'
export {default as Button} from './Button'
export {default as Icon} from './Icon'
export {default as Image} from './Image'
export {default as UserLogon} from './UserLogon'
export {default as Menu} from './Menu'
export {default as MenuItem} from './MenuItem'
export {default as NumberInput} from './NumberInput'
export {default as DateInput} from './DateInput'
export {default as TrueFalseInput} from './TrueFalseInput'
export {default as SpeechInput} from './SpeechInput'
export {default as Data} from './Data'
export {default as Calculation} from './Calculation'
export {default as Timer} from './Timer'
export {default as Collection} from './Collection'
export {default as Page} from './Page'
export {default as ComponentElement} from './ComponentElement'
export {default as Form} from './Form'
export {default as BaseFormState} from './FormState'
export {default as SelectInput} from './SelectInput'
export {default as TextElement} from './TextElement'
export {default as Layout} from './Layout'
export {default as Block} from './Block'
export {default as AppBar} from './AppBar'
export {default as ListElement} from './ListElement'
export {default as ListItem} from './ListItem'
export {default as ItemSet} from './ItemSet'
export {default as ItemSetItem} from './ItemSetItem'
export {default as TextInput} from './TextInput'
export {default as MemoryDataStore} from './MemoryDataStore'
export {default as FileDataStore} from './FileDataStore'
export {default as FileDataStoreImpl} from './FileDataStoreImpl'
export {default as IdbDataStoreImpl} from './IdbDataStoreImpl'
export {default as BrowserDataStore} from './BrowserDataStore'
export {default as FirestoreDataStore} from './FirestoreDataStore'
export {default as FirestoreDataStoreImpl} from './FirestoreDataStoreImpl'
export {default as ServerAppConnector} from './ServerAppConnector'
export {default as Adapter} from './Adapter'
