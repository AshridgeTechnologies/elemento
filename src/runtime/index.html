<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="initial-scale=1, width=device-width" />
  <title>Elemento App</title>
  <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,400,500,700&display=swap"/>
</head>
<body>
  <div id="main"></div>
  <script src="https://unpkg.com/react@17/umd/react.development.js" crossorigin></script>
  <script src="https://unpkg.com/react-dom@17/umd/react-dom.development.js" crossorigin></script>
  <script src="runApp.ts" type="module"></script>
</body>
</html>