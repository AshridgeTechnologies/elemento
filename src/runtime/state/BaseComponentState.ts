import {AppStateForObject, StoredState, StoredStateWithProps} from './AppStateStore'
import {equals} from 'ramda'

export class BaseComponentState<ExternalProps extends object, StateProps extends object = ExternalProps> implements StoredStateWithProps<ExternalProps> {
    protected state: StateProps = {} as StateProps
    protected _appStateInterface?: AppStateForObject<any>
    get _path(): string | undefined { return this._appStateInterface?.path }

    constructor(public props: ExternalProps) {}

    init(asi: AppStateForObject<any>, previousVersion?: this): this {
        this._appStateInterface = asi
        this.doInit(previousVersion)
        return this
    }

    protected doInit(_previousVersion?: this) {}

    updateState(changes: Partial<typeof this.state>) {
        const latestState = this.latest()
        const updatedState = latestState.withMergedState(changes)
        if (updatedState !== latestState) {
            this._appStateInterface!.update(updatedState)
        }
    }

    latest(): this {
        return this._appStateInterface!.latest() as this
    }

    withState(state: StateProps): this {
        return this.stateEqual(state) ? this : this.copy(this.props, state)
    }

    withProps(props: ExternalProps): this {
        return this.propsEqual(props) ? this : this.copy(props, this.state)
    }

    private withMergedState(changes: Partial<StateProps>): this {
        const newState = Object.assign({}, this.state, changes) as StateProps
        for (const p in newState) {
            if (newState[p] === undefined) delete newState[p]
        }

        return this.withState(newState)
    }

    get _stateForTest() {
        return this.state
    }

    protected copy(props: ExternalProps, state: StateProps): this {
        const ctor = this.constructor as new (props: ExternalProps) => this
        const newVersion = new ctor(props) as this
        newVersion.state = {...state}
        return newVersion
    }

    protected propsEqual(props: ExternalProps) {
        return equals(this.props, props)
    }

    protected stateEqual(state: StateProps) {
        return equals(this.state, state)
    }

    protected getChildState<T extends StoredState>(name: string): T {
        return this._appStateInterface?.getChildState(name) as T
    }
}

export function createProxy<T extends StoredState>(store: AppStateForObject<T>, targetState: T) {
    const handler = {
        get(target: T, property: string): any {
            if (property in target) {
                return target[property as keyof T]
            }

            return store.getChildState(property)
        }
    }

    return new Proxy(targetState, handler)
}

export class BaseComponentStateWithProxy<ExternalProps extends object, StateProps extends object = ExternalProps> extends BaseComponentState<ExternalProps, StateProps> {

    init(asi: AppStateForObject<any>, previousVersion?: this): this {
        return createProxy(asi, super.init(asi, previousVersion))
    }
}
