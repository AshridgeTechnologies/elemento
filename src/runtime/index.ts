import * as components from './components'
import {globalFunctions} from './globalFunctions'
export {components, globalFunctions}

export {default as React} from 'react'
export {default as ReactDOM} from 'react-dom/client'
export {default as appFunctions} from './appFunctions'
export {useGetObjectState, useObjectState} from './appData'
export {asArray, codeGenerationError, parentPath} from './runtimeFunctions'
export {default as AppRunner} from './AppRunner'
export {run} from './run'
