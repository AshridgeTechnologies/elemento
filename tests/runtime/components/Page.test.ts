/**
 * @vitest-environment jsdom
 */
import {expect, test} from 'vitest'
import React, {createElement} from 'react'
import {componentJSON, componentJSONAsync} from '../../testutil/testHelpers'
import TextElement from '../../../src/runtime/components/TextElement'
import {NumberInput, Page, TextInput} from '../../../src/runtime/components/index'
import {useComponentState} from '../../../src/runtime'
import {BaseComponentState} from '../../../src/runtime/state/BaseComponentState'


class TestPageState extends BaseComponentState<{}> {
}

function TestPage(props: {path: string} ) {
    const pathTo = (name: string) => props.path + '.' + name

    const _state = useComponentState(props.path, TestPageState, {})
    const {BoxSize} = _state as any

    return React.createElement(Page, props,
        React.createElement(TextInput, {path: pathTo('Description'), initialValue: 'Big', label: 'Description'}),
        React.createElement(NumberInput, {path: pathTo('BoxSize'), initialValue: 33, label: 'Size'}),
        // @ts-ignore
        React.createElement(TextElement, {path: pathTo('Feedback'), content: 'Size is ' + BoxSize} )
    )
}

test('Page element produces output containing children', () => {
    const child = createElement(TextElement, {path: 'page1.para1', styles: {maxWidth: 200}, content: 'Hello\nwhere are you'})
    const component = createElement(Page, {path: 'page1'}, child)
    expect(componentJSON(component)).toMatchSnapshot()
})

test('TestPage produces output with correct values from child elements', async () => {
    const component = createElement(TestPage, {path: 'page1'})
    expect(await componentJSONAsync(component)).toMatchSnapshot()
})
