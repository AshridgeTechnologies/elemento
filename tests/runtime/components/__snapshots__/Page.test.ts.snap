// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Page element produces output containing children 1`] = `
<div
  className="MuiStack-root ElPage css-pao7ie-MuiStack-root"
  id="page1"
  tabIndex={-1}
>
  <div
    className="MuiTypography-root MuiTypography-body1 css-1pnky8i-MuiTypography-root"
    id="page1.para1"
  >
    Hello
    <br />
    where are you
  </div>
</div>
`;
